gastime/gastime.go

```diff
@@ -0,0 +1,112 @@
+// [ACP-194]: https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/194-streaming-asynchronous-execution
+type Time struct {
+	TimeMarshaler
+}
+
+// makeTime is a constructor shared by [New] and [Time.Clone].
+func makeTime(t *proxytime.Time[gas.Gas], target, excess gas.Gas) *Time {
+	tm := &Time{
+		TimeMarshaler: TimeMarshaler{
+			Time:   t,
+			target: target,
+			excess: excess,
+		},
+	}
+	tm.establishInvariants()
+	return tm
+}
+
+func (tm *Time) establishInvariants() {
+	tm.Time.SetRateInvariants(&tm.target, &tm.excess)
+}
+
+// New returns a new [Time], set from a Unix timestamp. The consumption of
+// `2*target` units of [gas.Gas] is equivalent to a tick of 1 second.
+func New(unixSeconds uint64, target, startingExcess gas.Gas) *Time {
+	return makeTime(proxytime.New(unixSeconds, 2*target), target, startingExcess)
+}
+
+// Clone returns a deep copy of the time.
+func (tm *Time) Clone() *Time {
+	// [proxytime.Time.Clone] explicitly does NOT clone the rate invariants, so
+	// we reestablish them as if we were constructing a new instance.
+	return makeTime(tm.Time.Clone(), tm.target, tm.excess)
+}
+
+// Target returns the `T` parameter of ACP-176.
+func (tm *Time) Target() gas.Gas {
+	return tm.target
+}
+
+// Excess returns the `x` variable of ACP-176.
+func (tm *Time) Excess() gas.Gas {
+	return tm.excess
+}
+
+// Price returns the price of a unit of gas, i.e. the "base fee".
+func (tm *Time) Price() gas.Price {
+	return gas.CalculatePrice(1 /* M */, tm.excess, 87*tm.target /* K */)
+}
+
+// BaseFee is equivalent to [Time.Price], returning the result as a uint256 for
+// compatibility with geth/libevm objects.
+func (tm *Time) BaseFee() *uint256.Int {
+	return uint256.NewInt(uint64(tm.Price()))
+}
+
+// SetTarget changes the target gas consumption per second. It is equivalent to
+// [proxytime.Time.SetRate] with `2*t`, but is preferred as it avoids
+// accidentally setting an odd rate. It returns an error if the scaled
+// [Time.Excess] overflows as a result of the scaling.
+func (tm *Time) SetTarget(t gas.Gas) error {
+	_, err := tm.SetRate(2 * t) // also updates target as it was passed to [proxytime.Time.SetRateInvariants]
+	return err
+}
+
+// Tick is equivalent to [proxytime.Time.Tick] except that it also updates the
+// gas excess.
+func (tm *Time) Tick(g gas.Gas) {
+	tm.Time.Tick(g)
+
+	R, T := tm.Rate(), tm.Target()
+	quo, _, _ := intmath.MulDiv(g, R-T, R) // overflow is impossible as (R-T)/R < 1
+	tm.excess += quo
+}
+
+// FastForwardTo is equivalent to [proxytime.Time.FastForwardTo] except that it
+// may also update the gas excess.
+func (tm *Time) FastForwardTo(to uint64) {
+	sec, frac := tm.Time.FastForwardTo(to)
+	if sec == 0 && frac.Numerator == 0 {
+		return
+	}
+
+	R, T := tm.Rate(), tm.Target()
+	quo, _, _ := intmath.MulDiv(R*gas.Gas(sec)+frac.Numerator, T, R) // overflow is impossible as T/R < 1
```

Comment:
I agree that `MulDiv` will never return an error, but it does seem possible that `R * sec + Numerator` could overflow.

While that probably isn't relevant for C-chain, it will likely be relevant for subnet-evm